DoubleTap recognizer support and improved error message

GaryQian opened the pull request:
## Description

We currently only support `TapGestureRecognizer` and `LongPressGestureRecognizer` and `assert(false)` otherwise. This results in a very unclear error.

Add an error message to make this more clear.

We could support additional detectors, in particular, DoubleTap and MultiTap, if deemed appropriate.

## Related Issues

See https://github.com/flutter/flutter/issues/56304

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Did any tests fail when you ran them? Please read [Handling breaking changes].

- [x] No, no existing tests failed, so this is *not* a breaking change.
- [ ] Yes, this is a breaking change. *If not, delete the remainder of this section.*
   - [ ] I wrote a design doc: https://flutter.dev/go/template *Replace this with a link to your design doc's short link*
   - [ ] I got input from the developer relations team, specifically from: *Replace with the names of who gave advice*
   - [ ] I wrote a migration guide: *Replace with a link to your migration guide*

<!-- Links -->
[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

GaryQian committed:
Add error message to make it more clear what went wrong.:

GaryQian committed:
Remove if

GaryQian committed:
add is

jonahwilliams commented on packages/flutter/lib/src/rendering/paragraph.dart:
```diff
@@ -919,7 +919,7 @@ class RenderParagraph extends RenderBox
             configuration.onLongPress = recognizer.onLongPress;
           } else {
-            assert(false);
+            assert(false, '${recognizer.runtimeType} is not supported.');
```
Maybe: must be one of Tap/Longpress?

jonahwilliams replied:
Assets can be caught in unit tests too, just saying :)

GaryQian replied:
I didn't want to be too specific here, as I don't see why we cant support DoubleTap or MultiTap as well. If we add it shortly, then that would be overly specific and not strictly correct

jonahwilliams reviewed (approved):
LGTM

GaryQian committed:
Support double and multi tap

GaryQian committed:
Support DoubleTapGestureRecognizer

GaryQian commented:
I added DoubleTapGestureRecognizer since its onDoubleTap callback is mappable to the onTap configuration. Added tests to match. MultiTapGestureRecognizer's callback requires an int, so it is not directly usable as an onTap.
